src/hotspot/share/runtime/synchronizer.cpp

```diff
@@ -1591,119 +1566,184 @@ static size_t delete_monitors(GrowableArray<ObjectMonitor*>* delete_list) {
+
+  // The async deflation request has been processed.
+  _last_async_deflation_time_ns = os::javaTimeNanos();
+  set_is_async_deflation_requested(false);
+
+  ObjectMonitorDeflationLogging log;
+  ObjectMonitorDeflationSafepointer safepointer(&log);
+
+  log.begin();
+
   // Deflate some idle ObjectMonitors.
-  size_t deflated_count = deflate_monitor_list(current, ls, &timer);
+  size_t deflated_count = deflate_monitor_list(&safepointer);
+
+  // Unlink the deflated ObjectMonitros from the in-use list.
   size_t unlinked_count = 0;
   size_t deleted_count = 0;
   if (deflated_count > 0) {
-    // There are ObjectMonitors that have been deflated.
-
-    // Unlink deflated ObjectMonitors from the in-use list.
     ResourceMark rm;
     GrowableArray<ObjectMonitor*> delete_list((int)deflated_count);
-    unlinked_count = _in_use_list.unlink_deflated(current, ls, &timer, &delete_list);
-    if (current->is_monitor_deflation_thread()) {
-      if (ls != nullptr) {
-        timer.stop();
-        ls->print_cr("before handshaking: unlinked_count=" SIZE_FORMAT
-                     ", in_use_list stats: ceiling=" SIZE_FORMAT ", count="
-                     SIZE_FORMAT ", max=" SIZE_FORMAT,
-                     unlinked_count, in_use_list_ceiling(),
-                     _in_use_list.count(), _in_use_list.max());
-      }
+    unlinked_count = _in_use_list.unlink_deflated(&delete_list, &safepointer);
 
-      // A JavaThread needs to handshake in order to safely free the
-      // ObjectMonitors that were deflated in this cycle.
-      HandshakeForDeflation hfd_hc;
-      Handshake::execute(&hfd_hc);
-      // Also, we sync and desync GC threads around the handshake, so that they can
-      // safely read the mark-word and look-through to the object-monitor, without
-      // being afraid that the object-monitor is going away.
-      VM_RendezvousGCThreads sync_gc;
-      VMThread::execute(&sync_gc);
-
-      if (ls != nullptr) {
-        ls->print_cr("after handshaking: in_use_list stats: ceiling="
-                     SIZE_FORMAT ", count=" SIZE_FORMAT ", max=" SIZE_FORMAT,
-                     in_use_list_ceiling(), _in_use_list.count(), _in_use_list.max());
-        timer.start();
-      }
-    } else {
-      // This is not a monitor deflation thread.
-      // No handshake or rendezvous is needed when we are already at safepoint.
-      assert_at_safepoint();
-    }
+    log.before_handshake(unlinked_count);
+
+    // A JavaThread needs to handshake in order to safely free the
+    // ObjectMonitors that were deflated in this cycle.
+    HandshakeForDeflation hfd_hc;
+    Handshake::execute(&hfd_hc);
+    // Also, we sync and desync GC threads around the handshake, so that they can
+    // safely read the mark-word and look-through to the object-monitor, without
+    // being afraid that the object-monitor is going away.
+    VM_RendezvousGCThreads sync_gc;
+    VMThread::execute(&sync_gc);
```

Comment:
This is new to me. What does "sync and desync GC threads" mean?
And what does it mean that we didn't do this before? Is this a bug fix
that should be backported to older releases?